File: JAVA/21_Binary_Search/BOJ_03_1654_랜선자르기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int k = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());
        int[] arr = new int[k];
        long max = 0;
        for (int i = 0; i < k; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            max = Math.max(max, arr[i]);
        }
        long min = 1;
        while (min <= max) {
            long mid = (max + min) / 2;
            int cnt = 0;
            for (int i = 0; i < k; i++) {
                cnt += (arr[i] / mid);
            }
            if (cnt < n) {
                max = mid - 1;
            } else {
                min = mid + 1;
            }
        }
        System.out.println(max);
    }
}

File: C/21_Binary_Search/BOJ_03_1654_랜선자르기.c
#include <stdio.h>
#define math_max(a, b) a > b ? a : b
main() {
	int k, n;
	scanf("%d %d", &k, &n);
	int* arr = (int*)malloc(sizeof(int) * k);
	long long int max = 0;
	for (int i = 0; i < k; i++) {
		scanf("%d", &arr[i]);
		max = math_max(max, arr[i]);
	}
	long long int min = 1;
	while (min <= max) {
		long long int mid = (max + min) / 2;
		int cnt = 0;
		for (int i = 0; i < k; i++) {
			cnt += (arr[i] / mid);
		}
		if (cnt < n) {
			max = mid - 1;
		}
		else {
			min = mid + 1;
		}
	}
	printf("%d", max);
}